mypy/checkexpr.py

```diff
@@ -1104,68 +1106,242 @@ def check_arg(self, caller_type: Type, original_caller_type: Type,
                 if call:
                     self.msg.note_call(original_caller_type, call, context)
 
-    def overload_call_target(self, arg_types: List[Type], arg_kinds: List[int],
-                             arg_names: Optional[Sequence[Optional[str]]],
-                             overload: Overloaded, context: Context,
-                             messages: Optional[MessageBuilder] = None) -> Type:
-        """Infer the correct overload item to call with given argument types.
+    def check_overload_call(self,
+                            callee: Overloaded,
+                            args: List[Expression],
+                            arg_types: List[Type],
+                            arg_kinds: List[int],
+                            arg_names: Optional[Sequence[Optional[str]]],
+                            callable_name: Optional[str],
+                            object_type: Optional[Type],
+                            context: Context,
+                            arg_messages: MessageBuilder) -> Tuple[Type, Type]:
+        """Checks a call to an overloaded function."""
+        # Step 1: Filter call targets to remove ones where the argument counts don't match
+        plausible_targets = self.plausible_overload_call_targets(arg_types, arg_kinds,
+                                                                 arg_names, callee)
+
+        # Step 2: Attempt to find a matching overload
+        inferred_result = self.infer_overload_return_type(plausible_targets, args, arg_types,
+                                                          arg_kinds, arg_names, callable_name,
+                                                          object_type, context, arg_messages)
+        if inferred_result is not None:
+            # Success! Stop early.
+            return inferred_result
+
+        # Step 3: At this point, we know none of the overload alternatives exactly match.
+        #         We fall back to using the erased types to help with union math/help us
+        #         produce a better error message.
+        erased_targets = self.overload_erased_call_targets(plausible_targets, arg_types,
```

Comment:
Oh, I did it because if an overload has a signature of `(int) -> int` and `(List[int]) -> str`, and if a user tries passing in an argument of type `List[str]` or something, we probably want guess that the user meant the second option and give an error message based off of that.

(This also let us maintain compatibility with a lot of the existing error messages.)

I added a comment to clarify.